node/src/components/fetcher/trie_fetcher.rs

```diff
@@ -0,0 +1,342 @@
+            Event::TrieOrChunkFetched { id, .. } => write!(f, "trie or chunk {} fetched", id),
+        }
+    }
+}
+
+fn into_response<I>(trie: Box<Trie<Key, StoredValue>>, sender: Option<I>) -> TrieFetcherResult<I>
+where
+    I: Debug + Eq,
+{
+    match sender {
+        Some(peer) => Ok(FetchedData::FromPeer { item: trie, peer }),
+        None => Ok(FetchedData::FromStorage { item: trie }),
+    }
+}
+
+impl<I> TrieFetcher<I>
+where
+    I: Debug + Clone + Send + Eq + 'static,
+{
+    pub(crate) fn new() -> Self {
+        TrieFetcher {
+            partial_chunks: Default::default(),
+        }
+    }
+
+    fn consume_trie_or_chunk<REv>(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        sender: Option<I>,
+        trie_or_chunk: TrieOrChunkedData,
+    ) -> Effects<Event<I>>
+    where
+        REv: ReactorEventT<TrieOrChunkedData> + From<FetcherRequest<I, TrieOrChunkedData>>,
+    {
+        let TrieOrChunkedDataId(_index, hash) = trie_or_chunk.id();
+        match trie_or_chunk {
+            TrieOrChunkedData::Trie(trie) => match self.partial_chunks.remove(&hash) {
+                None => {
+                    error!(%hash, "fetched a trie we didn't request!");
+                    Effects::new()
+                }
+                Some(partial_chunks) => {
+                    debug!(%hash, "got a full trie");
+                    partial_chunks
+                        .responder
+                        .respond(into_response(trie, sender))
+                        .ignore()
+                }
+            },
+            TrieOrChunkedData::ChunkWithProof(chunk) => {
+                self.consume_chunk(effect_builder, sender, chunk)
+            }
+        }
+    }
+
+    fn consume_chunk<REv>(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        sender: Option<I>,
+        chunk: ChunkWithProof,
+    ) -> Effects<Event<I>>
+    where
+        REv: ReactorEventT<TrieOrChunkedData> + From<FetcherRequest<I, TrieOrChunkedData>>,
+    {
+        if !chunk.verify() {
+            match sender {
+                None => {
+                    error!(?chunk, "got an invalid chunk from storage");
+                    return Effects::new();
+                }
+                Some(sender) => {
+                    warn!(?sender, ?chunk, "got an invalid chunk from sender");
+                    // TODO: would be good to re-request from someone else instead of the same
+                    // node...
+                    let id = TrieOrChunkedDataId(chunk.proof().index(), chunk.proof().root_hash());
+                    return effect_builder
+                        .fetch(id, sender)
+                        .event(move |fetch_result| Event::TrieOrChunkFetched { id, fetch_result });
+                }
+            }
+        }
+        let digest = chunk.proof().root_hash();
+        let index = chunk.proof().index();
+        let count = chunk.proof().count();
```

Comment:
Isn't it insecure that it's the sender of the chunk decides how many more chunks are there to download? Sounds like an option to lead the receiving node into an infinite process of fetching the chunks.

Reply:
If we verify each chunk cryptographically they can't lie about that: The number of chunks is [included in the computation of the root hash](https://github.com/casper-network/casper-node/blob/f0c6601205a3e0db9139a94c12bc6095d7cf3ce0/hashing/src/indexed_merkle_proof.rs#L156).

Reply:
We can verify the root hash only after downloading all elements so while we're downloading separate chunks we cannot know how many are left(?). Maybe we should extend the process with that check though? I.e. we first download the "root" proof, that contains the count, and use that as the reference for when we are downloading the chunks and if the already-fetched count exceeds the one from the root then we fail.

Reply:
But we are calling `chunk.verify()` at the beginning of this method. If that passes we know that the count in that chunk is correct.

(@fizyk20: Actually, why not call that right away in `Item::verify`?)

Reply:
I believe that if the count is incorrect, then the chunk will not pass verification. IIRC, the count is included in the root hash. So if we requested a specific root hash, and it matches, the count has to be correct.

Reply:
And not in the `TrieFetcher` but implicitly in the `Fetcher` that calls `id()` when looking up the responder that requested the data 😵

Reply:
I commented here without refreshing and I only saw the first comment from @goral09 then :grimacing: So what's the consensus here? Should I move the call to `chunk.verify()` to `Item::verify`?